src/sabre_swap.rs

```diff
@@ -0,0 +1,138 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use ndarray::prelude::*;
+use numpy::PyReadonlyArray2;
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+use rayon::prelude::*;
+
+use crate::edge_list::EdgeList;
+use crate::nlayout::NLayout;
+use crate::qubits_decay::QubitsDecay;
+use crate::swap_scores::SwapScores;
+
+const EXTENDED_SET_WEIGHT: f64 = 0.5;
+
+#[pyclass]
+pub enum Heuristic {
+    Basic,
+    Lookahead,
+    Decay,
+}
+
+#[pyfunction]
+pub fn sabre_score_heuristic(
+    layer: EdgeList,
+    layout: &NLayout,
+    swap_scores: &mut SwapScores,
+    extended_set: EdgeList,
+    distance_matrix: PyReadonlyArray2<f64>,
+    qubits_decay: QubitsDecay,
+    heuristic: &Heuristic,
+) -> Vec<[usize; 2]> {
+    let dist = distance_matrix.as_array();
+    swap_scores
+        .scores
+        .par_iter_mut()
+        .for_each(|(swap_qubits, score)| {
+            let mut trial_layout = layout.clone();
+            trial_layout.swap_logic(swap_qubits[0], swap_qubits[1]);
+            *score = score_heuristic(
+                heuristic,
+                &layer.edges,
+                &extended_set.edges,
+                &trial_layout,
+                swap_qubits,
+                &dist,
+                &qubits_decay.decay,
+            );
+        });
+    let min_score = swap_scores
+        .scores
+        .par_values()
+        .min_by(|a, b| a.partial_cmp(b).unwrap())
+        .unwrap();
+    let mut best_swaps: Vec<[usize; 2]> = swap_scores
+        .scores
+        .iter()
+        .filter_map(|(k, v)| if v == min_score { Some(*k) } else { None })
+        .collect();
+    best_swaps.par_sort();
```

Comment:
Do we need to sort `best_swaps` if they're already all the same value?  I guess this is to do with keeping the same routing for the same random seed, but given the test changes, it looks like we already may be breaking something there.  That said, I think the differences in the tests might be to do with possible mistake in handling the decay mentioned above.

Reply:
Yeah, I only did this because it was in the python code. I assumed the sort was there mostly for repeatability because we're dependent on the insertion order otherwise. I can drop it though and see what happens, it would speed things up without needing to do this.